src/core/ext/filters/client_channel/lb_policy/xds/xds_load_balancer_api.cc

```diff
@@ -18,12 +18,129 @@
 
 #include <grpc/support/port_platform.h>
 
-#include "pb_decode.h"
-#include "pb_encode.h"
 #include "src/core/ext/filters/client_channel/lb_policy/xds/xds_load_balancer_api.h"
 
 #include <grpc/support/alloc.h>
 
+namespace grpc_core {
+
+namespace {
+
+constexpr char kEdsTypeUrl[] =
+    "type.googleapis.com/envoy.api.v2.ClusterLoadAssignment";
+constexpr char kEndpointRequired[] = "endpointRequired";
+upb_alloc* g_arena = nullptr;
+
+void InitArena() { g_arena = upb_arena_new(); }
+
+upb_alloc* arena() {
+  static gpr_once once = GPR_ONCE_INIT;
+  gpr_once_init(&once, InitArena);
+  return g_arena;
+}
+
+}  // namespace
+
+XdsDiscoveryRequest* XdsRequestCreate(const char* service_name) {
+  XdsDiscoveryRequest* request = envoy_api_v2_DiscoveryRequest_new(arena());
+  XdsNode* node = envoy_api_v2_DiscoveryRequest_mutable_node(request, arena());
+  XdsStruct* metadata = envoy_api_v2_core_Node_mutable_metadata(node, arena());
+  XdsFieldsEntry* field = google_protobuf_Struct_add_fields(metadata, arena());
+  google_protobuf_Struct_FieldsEntry_set_key(
+      field, upb_strview_make(kEndpointRequired, strlen(kEndpointRequired)));
+  XdsValue* value =
+      google_protobuf_Struct_FieldsEntry_mutable_value(field, arena());
+  google_protobuf_Value_set_bool_value(true);
+  envoy_api_v2_DiscoveryRequest_add_resource_names(request, service_name,
+                                                   arena());
+  envoy_api_v2_DiscoveryRequest_set_type_url(
+      request, upb_strview_make(kEdsTypeUrl, strlen(kEdsTypeUrl)));
+  return request;
+}
+
+grpc_resolved_address ServerAddressParse(const XdsLbEndpoint* lb_endpoint) {
+  grpc_resolved_address addr;
+  // FIXME
+  return addr;
+}
+
+LocalityUpdateArgs LocalityParse(const XdsLocalityLbEndpoints* locality) {
+  LocalityUpdateArgs update_args;
+  // Parse locality_name.
+  // FIXME: correct locality name.
+  upb_strview locality_name = envoy_api_v2_core_Locality_sub_zone(locality);
+  char* buf = static_cast<char*>(malloc(locality_name.size + 1));
+  memcpy(buf, locality_name.data, locality_name.size);
+  buf[locality_name.size] = '\0';
+  update_args.locality_name = UniquePtr<char>(buf);
+  // Parse the addresses.
+  size_t size;
+  XdsLbEndpoint** lb_endpoints =
+      envoy_api_v2_endpoint_LocalityLbEndpoints_lb_endpoints(locality, &size);
+  for (size_t i = 0; i < size; ++i) {
+    update_args.addresses.emplace_back(ServerAddressParse(lb_endpoints[i]),
+                                       nullptr);
+  }
+  // Parse the lb_weight and priority.
+  const google_protobuf_UInt32Value* lb_weight =
+      envoy_api_v2_endpoint_LbEndpoint_load_balancing_weight(locality);
+  update_args.lb_weight = google_protobuf_UInt32Value_value(lb_weight);
```

Comment:
lb_weight will be NULL if the field was unset.

Reply:
Good point! I will add sanity check.